### ISSUE TITLE
Question about Flux SafeGuider recognizer (Flux model variant + embedding dimension)

### ISSUE THREAD

**Issue by DashKev** [pgqihere/SafeGuider#4]
Hi authors,

Thanks for releasing SafeGuider and the pre-trained recognizer weights. I’m trying to reproduce the Flux-side SafeGuider pipeline and ran into a dimension mismatch that I’d like to clarify.

**What I observed**
- You provide `Models/Flux_safeguider.pt`. Its first layer has `fc1.weight` shape `(1024, 1024)`, so it expects a 1024-d input embedding.
- In my setup with HuggingFace diffusers `FluxPipeline` (I tested `black-forest-labs/FLUX.1-dev`), the T5 text encoder (`text_encoder_2`) has `d_model = 4096` and the per-token hidden state for the `</s>` token is 4096-d.
- As a result, feeding the `</s>` embedding from `text_encoder_2` directly into `Flux_safeguider.pt` fails due to `4096 != 1024`.

**Questions**
1) Which Flux model variant/config did you use for training and evaluation of `Flux_safeguider.pt`?
   - FLUX.1-dev or FLUX.1-schnell (or another checkpoint)?
2) For the Flux recognizer input, do you use:
   - the T5 encoder `</s>` token embedding from `text_encoder_2`, or
   - a projected/pooled embedding (e.g., a learned projection from 4096→1024), or
   - the CLIP text encoder output (768-d) instead of T5?
3) If a projection is involved, could you share:
   - the exact projection method/code (e.g., linear layer, which weights, where to load it), or
   - the exact feature definition that results in a 1024-d input?
4) Could you also share the recommended inference code path for Flux (since the repository mainly provides the SD1.4 end-to-end generation script)?

Any clarification would be greatly appreciated—especially the exact Flux checkpoint (dev vs schnell) and the exact embedding extraction used for `Flux_safeguider.pt`.

Best regards,  
DashKev

**Comment by pgqihere** [comment]
Hi DashKev,

Thanks for your interest and the detailed question. Let me clarify the Flux-side pipeline.

## Regarding the Dimension Mismatch

You're correct that the T5 encoder (`text_encoder_2`) in FLUX.1-dev produces **4096-d** hidden states, while `Flux_safeguider.pt` expects **1024-d** input. This is by design — we apply a **fixed linear projection from 4096 → 1024** before feeding the embedding into the classifier.

Specifically, we use a deterministic linear layer initialized with `xavier_uniform_` under a fixed seed (`torch.manual_seed(42)`) and with zero bias. This layer's weights are **frozen** (no gradient), so it acts as a fixed, reproducible dimensionality reduction rather than a learned projection. As long as you use the same seed and initialization scheme, the projection is fully deterministic and you'll get matching results without needing to load separate projection weights.

---

## Answers to Your Questions

### Q1 – Flux Variant

We used **FLUX.1-dev** for both training and evaluation of `Flux_safeguider.pt`.

### Q2 – Embedding Source

We use the **T5 encoder `</s>` token embedding** from `text_encoder_2`, not CLIP. Concretely, we locate the position where `input_ids == 1` (the EOS/`</s>` token) and extract the hidden state at that position, then project it from 4096-d to 1024-d via the fixed linear layer described above.

### Q3 – Projection Method

The projection is a single `nn.Linear(4096, 1024)` with frozen weights. The key code is:

```python
class FixedDimensionReducer(nn.Module):
    def __init__(self, input_dim, output_dim):
        super().__init__()
        self.linear = nn.Linear(input_dim, output_dim)
        with torch.no_grad():
            torch.manual_seed(42)
            nn.init.xavier_uniform_(self.linear.weight)
            nn.init.zeros_(self.linear.bias)
            self.linear.weight.requires_grad = False
            self.linear.bias.requires_grad = False

    def forward(self, x):
        return self.linear(x)
```

No separate weight file is needed — the deterministic seed ensures reproducibility.

### Q4 – Inference Path for Flux

The overall inference flow is:

1. Load the T5 encoder and tokenizer from `FLUX.1-dev/text_encoder_2` and `FLUX.1-dev/tokenizer_2`.
2. Tokenize the prompt and extract the T5 hidden state at the `</s>` token position.
3. Apply `FixedDimensionReducer(4096, 1024)` to get a 1024-d embedding.
4. Feed the reduced embedding into `Flux_safeguider.pt` (a three-layer classifier: `1024 → 1024 → 512 → 2`).
5. The softmax output gives `[unsafe_prob, safe_prob]`; `argmax` yields the prediction (`0 = unsafe`, `1 = safe`).

---

## A Note on the Two Available Approaches

We support two modes for SafeGuider:

- **Plug-and-play with CLIP:** For architectures that use a CLIP text encoder (e.g., SD 1.x), you can directly use the CLIP `</s>` token embedding (768-d) with the corresponding recognizer weights. This requires no projection step and works out of the box.
- **T5-based classifier (Flux):** Since Flux relies on T5 rather than CLIP, the Flux-side recognizer is trained on T5 embeddings with the fixed linear projection to bridge the dimension gap. The `.pt` file you have (`Flux_safeguider.pt`) is this T5-based variant.

Thanks again for raising this — it's a good catch and will help other users as well.

Best regards
